### PR TITLE
add volume fraction functions

### PR DESCRIPTION
**Pull request by cpgoodri**
This PR adds functionality to calculate volume fractions in arbitrary dimensions. Specifically, it adds the following 4 functions:

- `nball_unit_volume`: calculate the volume of an n-dimensional unit sphere
- `particle_volume`: calculate the total volume of a collection of particles. There are various options for specifying different particle radii (e.g. species, etc.). 
-  `volume_fraction`: calculate the volume fraction, uses `quantity.volume`
- `box_size_at_volume_fraction`: similar to `box_size_at_number_density` but for volume fraction. 

While I did document these functions, a few examples might be useful to explain how to use the parameters `radii`, `particle_count`, and `species` to cover different situations:

- case 1: monodisperse particles

```python
dim = 2
phi = 0.87
radii = 1
N = 100
print(particle_volume(radii, dim, particle_count = N))
box_size = box_size_at_volume_fraction(phi, radii, dim, particle_count = N)
print(box_size)
print(volume_fraction(box_size, radii, dim, particle_count = N))
```
- case 2: specify each radii separately
```python
radii = jnp.linspace(1,1.4,N)
print(particle_volume(radii, dim))
box_size = box_size_at_volume_fraction(phi, radii, dim)
print(box_size)
print(volume_fraction(box_size, radii, dim))
```
- case 3: species but with equal concentrations
```python
radii = jnp.array([1,1.1,1.2,1.3,1.4])
particle_count = N // radii.shape[0] #not particle_count = N !
print(particle_volume(radii, dim, particle_count = particle_count))
box_size = box_size_at_volume_fraction(phi, radii, dim, particle_count = particle_count)
print(box_size)
print(volume_fraction(box_size, radii, dim, particle_count = particle_count))
```
- case 4: species with unequal concentrations
```python 
radii = jnp.array([1,1.1,1.2,1.3,1.4])
particle_count = jnp.array([18,19,20,21,22])
print(particle_volume(radii, dim, particle_count = particle_count))
box_size = box_size_at_volume_fraction(phi, radii, dim, particle_count = particle_count)
print(box_size)
print(volume_fraction(box_size, radii, dim, particle_count = particle_count))
```
- case 5: using a jax-md species array
```python 
radii = jnp.array([1,1.1,1.2,1.3,1.4])
species = random.randint(random.PRNGKey(0), (N,), 0, radii.shape[0])
print(particle_volume(radii, dim, species = species))
box_size = box_size_at_volume_fraction(phi, radii, dim, species = species)
print(box_size)
print(volume_fraction(box_size, radii, dim, species = species))
```

Note: the case 5 cannot be jitted because the size of `jnp.bincount(species)` cannot be determined. This should be fairly easy to work around in practice by manually doing the bincount and then using case 4. 

As always, I'm eager to hear any comments/suggestions.

### CODE REVIEW

File: jax_md/quantity.py [hunk 1]
```diff
@@ -459,6 +459,56 @@ def g_fn(R, neighbor):
...
+    radii: array of shape (n,) giving particle radii
+    spatial_dimension: int giving the spatial dimension
+    particle_count: number of particles with each radii. broadcastable to radii.
+    species: list of particle species. If provided, this overrides 
+      particle_count.
+  
+  Returns: the sum of the volume of all the particles
+  """
+  V_unit = nball_unit_volume(spatial_dimension)
+  V_particle = V_unit * jnp.power(radii, spatial_dimension)
```
Comment (sschoenholz) [inline jax_md/quantity.py:482]: Can just be `radii**spatial_dimension`.

File: jax_md/quantity.py [hunk 2]
```diff
@@ -459,6 +459,56 @@ def g_fn(R, neighbor):
       return g_R
   return neighbor_fn, g_fn
 
+def nball_unit_volume(spatial_dimension: int) -> float:
```
Comment (sschoenholz) [inline jax_md/quantity.py:462]: Any chance of adding a comment here?

File: jax_md/quantity.py [hunk 3]
```diff
@@ -459,6 +459,56 @@ def g_fn(R, neighbor):
...
+  return jnp.power(jnp.pi, spatial_dimension / 2) / \
+    jnp.exp( jsp.special.gammaln(spatial_dimension / 2 + 1))
+
+def particle_volume(radii: Array,
```
Comment (sschoenholz) [inline jax_md/quantity.py:466]: If the user supplies species, would we expect per-particle radii or per-species? I sort of naively would expect the latter (or for something that mimics the convention of `smap` in this regard). But just glancing at the code (and comments) it always seems like this function requires a flat array of length num atoms?

**Review by sschoenholz — COMMENTED** [review]
Looking great, added a few minor comments then I'll commit.

File: jax_md/quantity.py (hunk 3 above)
Comment (cpgoodri) [reply jax_md/quantity.py:466]: The only time radii is should have size num atoms is in the "case 2" example above. In cases 3-5 (including whenever species is specified) radii is interpreted as per-species. I think the confusion was in the documentation for radii, which says it should have shape `(n,)`... but this `n` doesn't have to be the number of particles. I changed the documentation to hopefully make this more clear. Perhaps I should also add the 5 examples?

**Comment by cpgoodri** [comment]
Updated. Let me know if you think the documentation is clear enough, especially regarding the different use cases. Would it make sense to include those 5 example cases in the documentation?

**Pull request merged**